src/EnergyPlus/PhaseChangeModeling/HysteresisModel.cc

```diff
@@ -0,0 +1,344 @@
+				this->enthalpyM = this->getEnthalpy(updatedTempTDT, this->peakTempMelting, this->deltaTempMeltingLow, this->deltaTempMeltingHigh );
+				this->enthalpyF = this->getEnthalpy(updatedTempTDT, this->peakTempFreezing, this->deltaTempFreezingLow, this->deltaTempFreezingHigh );
+				if ( this->enthNew < this->enthOld && updatedTempTDT < prevTempTD) {
+					phaseChangeState = 0;
+					this->enthNew = ( this->specHeatTransition * updatedTempTDT ) + ( this->enthOld - ( this->specHeatTransition * prevTempTD) );
+				} else if ( this->enthNew < this->enthalpyF && this->enthNew > this->enthalpyM && updatedTempTDT < prevTempTD) {
+					phaseChangeState = 0;
+					this->enthNew = ( this->specHeatTransition * updatedTempTDT ) + ( this->enthRev - ( this->specHeatTransition * phaseChangeTempReverse ) );
+				} else if ( this->enthNew >= this->enthalpyF && updatedTempTDT <= phaseChangeTempReverse ) {
+					phaseChangeState = 0;
+					this->enthNew = ( this->specHeatTransition * updatedTempTDT ) + ( this->enthRev - ( this->specHeatTransition * phaseChangeTempReverse ) );
+				}
+			} else if ( prevPhaseChangeState == PhaseChangeStates::TRANSITION && phaseChangeState == PhaseChangeStates::FREEZING ) {
+				this->enthalpyM = this->getEnthalpy(updatedTempTDT, this->peakTempMelting, this->deltaTempMeltingLow, this->deltaTempMeltingHigh );
+				this->enthalpyF = this->getEnthalpy(updatedTempTDT, this->peakTempFreezing, this->deltaTempFreezingLow, this->deltaTempFreezingHigh );
+				this->enthRev = this->getEnthalpy(phaseChangeTempReverse, this->peakTempFreezing, this->deltaTempFreezingLow, this->deltaTempFreezingHigh );
+				this->enthNew = ( this->specHeatTransition * updatedTempTDT ) + ( this->enthRev - ( this->specHeatTransition * phaseChangeTempReverse ) );
+			}
+		}
+
+		// then calculate the specific heat and return it
+		if ( !this->phaseChangeTransition ) {
+			if ( this->enthNew == this->enthOld ) {
+				Cp = this->CpOld;
+			} else {
+				Cp = this->specHeat( prevTempTD, updatedTempTDT, Tc, Tau1, Tau2, this->enthOld, this->enthNew );
+			}
+		} else {
+			Cp = this->specHeatTransition;
+		}
+		this->CpOld = Cp;
+		return Cp;
+	}
+
+	Real64 HysteresisPhaseChange::specHeat( Real64 temperaturePrev, Real64 temperatureCurrent, Real64 criticalTemperature, Real64 tau1, Real64 tau2, Real64 EnthalpyOld, Real64 EnthalpyNew ) {
+
+		//	Tc                  ! Critical (Melting/Freezing) Temperature of PCM
+		//	Tau1                ! Width of Melting Zone low
+		//	Tau2                ! Width of Melting Zone high
+		//	EnthalpyOld         ! Previos Timestep Nodal Enthalpy
+		//	EnthalpyNew         ! Current Timestep Nodal Enthalpy
+
+        Real64 Cp1 = this->specificHeatSolid;
+        Real64 Cp2 = this->specificHeatLiquid;
+        Real64 T = temperatureCurrent;
+
+        Real64 DEta1 = - ( this->totalLatentHeat * ( T - criticalTemperature ) * exp( -2 * abs( T - criticalTemperature ) / tau1 ) ) / ( tau1 * abs( T - criticalTemperature ) );
```

Comment:
Reworked that function so that the intermediate variables are only calculated when valid, good to go here now.